hack/role/role.go

```diff
@@ -0,0 +1,361 @@
+		{
+			"Cluster Ingress Operator",
+			"0000_50_cluster-ingress-operator_00-ingress-credentials-request.yaml",
+			"0336e1d3-7a87-462b-b6db-342b63f7802c",
+		},
+		{
+			"Cluster Network Operator",
+			"0000_50_cluster-network-operator_02-cncc-credentials.yaml",
+			"be7a6435-15ae-4171-8f30-4a343eff9e8f",
+		},
+		{
+			"Cluster Storage Operator",
+			"0000_50_cluster-storage-operator_03_credentials_request_azure.yaml",
+			"5b7237c5-45e1-49d6-bc18-a1f62f400748",
+		},
+		{
+			"Cluster Storage Operator (Azure File)",
+			"0000_50_cluster-storage-operator_03_credentials_request_azure_file.yaml",
+			"0d7aedc0-15fd-4a67-a412-efad370c947e",
+		},
+	}
+)
+
+func main() {
+	flag.Parse()
+	if *verifiedVersion == "" {
+		panic("verified-version is required")
+	}
+
+	ctx := context.Background()
+
+	logger := logrus.New()
+	logger.SetLevel(logrus.ErrorLevel)
+	log := logrus.NewEntry(logger)
+	environment, err := env.NewCoreForCI(ctx, log) // we don't use log here
+	if err != nil {
+		panic(err)
+	}
+
+	options := environment.Environment().EnvironmentCredentialOptions()
+	tokenCredential, err := azidentity.NewEnvironmentCredential(options)
+	if err != nil {
+		panic(err)
+	}
+	roleDefinitions, err := armauthorization.NewArmRoleDefinitionsClient(tokenCredential, nil)
+	if err != nil {
+		panic(err)
+	}
+
+	m := manager{
+		roleDefinitions: roleDefinitions,
+	}
+
+	verifiedDir, err := os.MkdirTemp("", *verifiedVersion)
+	if err != nil {
+		panic(err)
+	}
+	err = extractCredReq(*verifiedVersion, verifiedDir)
+	if err != nil {
+		panic(err)
+	}
+
+	var vers []string
+	if *targetVersion != "" {
+		// if target version is specified, validate only that version
+		vers = []string{*targetVersion}
+	} else {
+		// validate all available versions
+		vers, err = versionsToValidate(*verifiedVersion)
+		if err != nil {
+			panic(err)
+		}
+	}
+
+	var missing []string
+	for _, v := range vers {
+		fmt.Println("Checking", v)
+		targetDir, err := os.MkdirTemp("", v)
+		if err != nil {
+			panic(err)
+		}
+		err = extractCredReq(v, targetDir)
+		if err != nil {
+			panic(err)
+		}
+
+		if err = validate(ctx, &m, verifiedDir, targetDir, permissionSets); err != nil {
+			// To check all versions, we need to continue even if there are missing permissions.
+			if errors.Is(err, &missingPermissionsError{}) {
+				missing = append(missing, v)
+				continue
+			}
+			panic(err)
+		}
+	}
+	if len(missing) > 0 {
+		for _, v := range missing {
+			fmt.Printf("Permissions are missing in %s\n", v)
+		}
+		os.Exit(1)
+	}
+}
+
+// versionsToValidate returns available versions later than the verified version
+func versionsToValidate(verifiedVersion string) ([]string, error) {
+	v, err := version.ParseVersion(verifiedVersion)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse version: %w", err)
+	}
+
+	var vers []string
+	for v.V[1]++; ; v.V[1]++ {
+		res, err := http.Get(fmt.Sprintf("https://api.openshift.com/api/upgrades_info/v1/graph?channel=fast-%s", v.MinorVersion()))
```

Comment:
I used the fast channel because I don't know how to handle candidate versions like 4.17.0-rc.0 or 4.17.0-ec.3.
We can use the candidate channel if we can skip those versions.

Reply:
I changed to use candidate channels.